File: backend/app/api/v1/routes/course_feedback.py
from fastapi import APIRouter, Depends, Header, Query, Request

from app.core.auth import get_current_actor, get_evaluator, verify_origin
from app.models.common import ApiResponse
from app.models.course_feedback import (
    CourseApplicationRegrade,
    CourseCorrectionCreate,
    CourseCorrectionReview,
    CourseCorrectionView,
    CourseFeedbackCreate,
    CourseFeedbackList,
    CourseFeedbackUpdate,
    CourseFeedbackView,
)
from app.services import course_feedback_service as service

router = APIRouter(prefix="/courses", tags=["course-feedback"])


@router.post("/{course_id}/feedback", status_code=201, response_model=ApiResponse[CourseFeedbackView])
async def create_feedback(
    course_id: str,
    body: CourseFeedbackCreate,
    actor=Depends(get_current_actor),
    idempotency_key: str = Header(..., min_length=1, max_length=128),
):
    return ApiResponse.success(await service.create_course_feedback(actor, course_id, body, idempotency_key))


@router.get("/{course_id}/feedback", response_model=ApiResponse[CourseFeedbackList])
async def list_feedback(
    course_id: str,
    actor=Depends(get_current_actor),
    lesson_id: str | None = Query(default=None),
    check_attempt_id: str | None = Query(default=None),
    course_assessment_id: str | None = Query(default=None),
):
    return ApiResponse.success(
        await service.list_course_feedback(
            actor.owner_id,
            course_id,
            lesson_id=lesson_id,
            check_attempt_id=check_attempt_id,
            course_assessment_id=course_assessment_id,
        )
    )


@router.patch("/{course_id}/feedback/{feedback_id}", response_model=ApiResponse[CourseFeedbackView])
async def update_feedback(
    course_id: str,
    feedback_id: str,
    body: CourseFeedbackUpdate,
    actor=Depends(get_current_actor),
):
    return ApiResponse.success(await service.update_course_feedback(actor.owner_id, course_id, feedback_id, body))


@router.post(
    "/{course_id}/feedback/{feedback_id}/corrections",
    status_code=201,
    response_model=ApiResponse[CourseCorrectionView],
)
async def create_correction(
    course_id: str,
    feedback_id: str,
    body: CourseCorrectionCreate,
    actor=Depends(get_current_actor),
    idempotency_key: str = Header(..., min_length=1, max_length=128),
):
    return ApiResponse.success(
        await service.create_correction(actor.owner_id, course_id, feedback_id, body, idempotency_key)
    )


@router.post(
    "/{course_id}/feedback/{feedback_id}/reviews",
    response_model=ApiResponse[CourseCorrectionView],
)
async def review_correction(
    course_id: str,
    feedback_id: str,
    body: CourseCorrectionReview,
    request: Request,
    evaluator=Depends(get_evaluator),
):
    from app.core.auth import verify_origin

    verify_origin(request)
    return ApiResponse.success(
        await service.review_correction(evaluator, course_id, feedback_id, body)
    )


@router.post(
    "/{course_id}/assessments/{course_assessment_id}/attempts/{attempt_id}/reviews",
    response_model=ApiResponse[dict],
)
async def review_application_attempt(
    course_id: str,
    course_assessment_id: str,
    attempt_id: str,
    body: CourseApplicationRegrade,
    request: Request,
    evaluator=Depends(get_evaluator),
):
    from app.core.auth import verify_origin

    verify_origin(request)
    return ApiResponse.success(
        await service.review_application_attempt(
            evaluator, course_id, course_assessment_id, attempt_id, body
        )
    )

File: web/src/services/courseFeedback.ts
import { request } from './http'
import type { ApiSchemas } from '../types/api'

export type CourseFeedbackCreate = ApiSchemas['CourseFeedbackCreate']
export type CourseFeedbackView = ApiSchemas['CourseFeedbackView']
export type CourseFeedbackList = ApiSchemas['CourseFeedbackList']
export type CourseFeedbackUpdate = ApiSchemas['CourseFeedbackUpdate']
export type CourseCorrectionView = ApiSchemas['CourseCorrectionView']
export type CourseFeedbackTarget = CourseFeedbackCreate['target']

const segment = encodeURIComponent

/** 保存问题；同 key 重放返回原记录，异 body 409。 */
export const createCourseFeedback = (
  courseId: string,
  body: CourseFeedbackCreate,
  key: string,
  signal?: AbortSignal,
) =>
  request<CourseFeedbackView>(`/courses/${segment(courseId)}/feedback`, {
    method: 'POST',
    data: body,
    idempotencyKey: key,
    signal,
  })

export const listCourseFeedback = (
  courseId: string,
  query: { lesson_id?: string; check_attempt_id?: string; course_assessment_id?: string } = {},
  signal?: AbortSignal,
) => {
  const params = new URLSearchParams()
  if (query.lesson_id) params.set('lesson_id', query.lesson_id)
  if (query.check_attempt_id) params.set('check_attempt_id', query.check_attempt_id)
  if (query.course_assessment_id) params.set('course_assessment_id', query.course_assessment_id)
  const suffix = params.size ? `?${params.toString()}` : ''
  return request<CourseFeedbackList>(`/courses/${segment(courseId)}/feedback${suffix}`, { signal })
}

export const updateCourseFeedback = (
  courseId: string,
  feedbackId: string,
  body: CourseFeedbackUpdate,
  signal?: AbortSignal,
) =>
  request<CourseFeedbackView>(
    `/courses/${segment(courseId)}/feedback/${segment(feedbackId)}`,
    { method: 'PATCH', data: body, signal },
  )

/** 个人纠正：append-only；provisional，不修改旧记录。 */
export const createCourseCorrection = (
  courseId: string,
  feedbackId: string,
  body: ApiSchemas['CourseCorrectionCreate'],
  key: string,
  signal?: AbortSignal,
) =>
  request<CourseCorrectionView>(
    `/courses/${segment(courseId)}/feedback/${segment(feedbackId)}/corrections`,
    { method: 'POST', data: body, idempotencyKey: key, signal },
  )

export const confirmedCorrections = (view: CourseFeedbackView) =>
  (view.corrections ?? []).filter((item) => item.confirmation === 'confirmed')

export const latestCorrection = (view: CourseFeedbackView) =>
  view.corrections?.length ? view.corrections[view.corrections.length - 1] : null

/** 一个人工确认能改变什么：注释、替代判分、请求新检查或准备内容修订。 */
export const correctionEffectLabel = (view: CourseFeedbackView): string => {
  const confirmed = confirmedCorrections(view)
  if (!confirmed.length) return '尚无已确认纠正，只有你自己的说明。'
  if (view.target.kind === 'course_assessment') return '已确认复核替代了原判分。'
  if (view.target.kind === 'quiz') return '已确认题目有误，原成绩保留，稍后安排新的检查。'
  return '已确认内容需要修订，正文更新前会先给你预览。'
}
